projects/ngx-clipboard/src/lib/ngx-clipboard.module.ts

```diff
@@ -1,12 +1,26 @@
 import { CommonModule } from '@angular/common';
-import { NgModule } from '@angular/core';
+import { ModuleWithProviders, NgModule } from '@angular/core';
 
-import { ClipboardDirective } from './ngx-clipboard.directive';
+import { CLIPBOARD_ROOT_OPTIONS } from './injection-tokens';
+import { ClipboardParams } from './interface';
 import { ClipboardIfSupportedDirective } from './ngx-clipboard-if-supported.directive';
+import { ClipboardDirective } from './ngx-clipboard.directive';
+import { ClipboardService } from './ngx-clipboard.service';
 
 @NgModule({
     imports: [CommonModule],
     declarations: [ClipboardDirective, ClipboardIfSupportedDirective],
-    exports: [ClipboardDirective, ClipboardIfSupportedDirective]
+    exports: [ClipboardDirective, ClipboardIfSupportedDirective],
+    providers: [ClipboardService]
 })
-export class ClipboardModule {}
+export class ClipboardModule {
+    static forRoot(params: ClipboardParams): ModuleWithProviders {
```

Comment:
Adding a special ID to the textarea seems hacky. I'll add the method to the service. However, I'd like to also propose a small architectural change. The current ClipboardService contains methods that are expected to be used by the consumers of the lib (such as `copyFromContent`) as well as methods that are used by the lib itself internally.

I think it'd be good to separate these out to make a clear public API service and an internal service that's not exposed to consumers.
The current `ClipboardService ` would be renamed into `ClipboardInternalService` and stay unchanged. The public service would have the name of the old service.

```ts
// that's new public API service. Notice - it only has two methods
class ClipboardService {
  constructor(private clipboard: ClipboardInternalService) {}

  configure(config: Config) {
    // .. set internal service configuration - basically what forRoot currently does
  }

  copyFromContent(content: string, container: HTMLElement = this.window.document.body) {
    this.clipboard.copyFromContent(content, container);
  }
}
```

That would be potentially a breaking change since someone out there might be calling the methods we would me making internal, but I'd argue that these methods should have been internal from the get go. It's just the TypeScript does not have such construct.

What do you think?

Reply:
I agree, your solution is more strict. However, I think using public and private to separate them is probably enough for this size of project. 

If you think it is a must, we can leave it in a different pull request and do it with next breaking change, which probably will happen in Angular 9.

Reply:
it's not a must. Just trying to make things cleaner. I'll leave it out

Reply:
I agree, if you are still interested to make the change, welcome to submit another pull-request around angular 9. I am sure there will be a breaking change at that time as well.